Introduce Reliability network layer

v0d1ch opened the pull request:
:sunflower:  This PR aims to improve our network stack and make it more resilient by using message tracking and resending missed messages. 

❓ This PR only addresses part of the resilience story from #188, eg. the part that deals with connection failures. Future PR should address the other issues, eg. resilience to transient node crashes

🚧 TODO:

* [x] test it in _real life_ with the demo (shut down the network for one node, post a transaction, restart the network, it should work)
* [x] Move ADR to another PR 
* [x] Write stress/model-based tests to improve confidence in the solution _(optional)_
* [x] More tests coverage for potential failures
     * remove use of partial functions, eg. `fromJust` from the code
     * have unit tests for each error path
* [x] ~Prune messages backlog list and limit its growth as we probably don't want it to clog memory~
     * Turns out this feature needs some more ❤️ to be ready for prime time so we'll postpone it to work on #1079 
* [x] Fix Log schemas

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed: 
     * Actually we want to update https://hydra.family/head-protocol/core-concepts/architecture/networking  page which is outdated, but this will be done in a later PR
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

pgrange committed:
WIP initiating a network model

pgrange committed:
WIP trying to model the network

v0d1ch committed:
Move type errors further

pgrange committed:
Better naming

pgrange committed:
First working skeleton

pgrange committed:
WIP ensuring some stuff

pgrange committed:
WIP try to express a property

pgrange committed:
SOme experiment

abailly committed:
Introduce message id for reliability layer

v0d1ch committed:
Introduce Authenticated messages

abailly committed:
Track each peer message index

pgrange committed:
Test message filtering as a property

v0d1ch committed:
Keep a list of last seen message ids

abailly committed:
Implement resending of unacked messages

This code is just an experiment and would require a lot of masssaging
to get right...

abailly committed:
Wire Reliability network layer into hydra-node

abailly committed:
Convert resent messages test to a property

abailly committed:
Allow network component to handle different in/out messages

abailly committed:
Check content of messages resent

abailly committed:
Check resent messages are consistent with original ones

abailly committed:
Add some logging to reliability layer

v0d1ch committed:
Use one map to track all ack messages

- Add a test case to make sure the acks are updated on each
  received/broadcasted message

pgrange committed:
Minor fixes/clarifications

abailly committed:
Split heartbeat incoming and outgoing network layers

abailly committed:
Added some comments to Reliability layer

v0d1ch committed:
Use Vectors instead of Lists

To be more performant let's use Vectors in the Msg type
as well as trace messages.

v0d1ch committed:
Add helper for constructing new acks

Rename some binders to make more sense

v0d1ch committed:
Use Vector as the main data structure

- We used Sequence to capture messages and List to hold parties.
Our benchmarks show worse performance than current master (they are not
able to complete at all). By using Vector for everything I am noticing
better performance but benchmarks for hydra-cluster stil don't finish.

- Also introduce custom exception types  for Reliability

v0d1ch committed:
Re-read acks upon sending

After possible callback local acks might change so we need to re-read
them from a TVar.

Add a party in the BroadcastCounter to improve logging.

abailly committed:
Simplify traces for Reliability

abailly committed:
Ensure changes to vector clock are atomic

We had an issue whereby concurrent call to the callback function from
lower layer lead to vector clock losing sync and not allowing progress
anymore because observation and changes were not atomic.

The resending logic does not seem to much make sense even though it
seems to work, we need to analyse the behaviour of the system a bit
more...

abailly committed:
Simplify authentication layer to not require passing a wrapped message

The Authenticate layer has the signing key to sign messages, so there is no
need to require senders to do the work themselves. This is possible because
we have decoupled incoming from outgoing messages in the definition of a
NetworkComponent.

abailly committed:
Wire heartbeat into reliability

The idea is that the reliability layer works with heartbeats to send
acks which do not change upon a ping, denoting the fact the node
cannot make progress by sending new messages and giving the
opportunity to peers to resend messages not acknowledged because of
transient network failures.

This makes the Reliability highly dependent on Heartbeat which raises
the question of whether or not we would not like to merge the 2.

abailly committed:
Randomly drop messages to simulate network failures

This is of course not meant to stay but is an interesting small
technique that I thought would be nice to introduce. A similar
technique could be used to simulate messages shuffling, or delays,
etc. Next commit will revert this.

abailly committed:
Revert "Randomly drop messages to simulate network failures"

This reverts commit 86dbd743a3735804bea0220bd140dd7ca1670908.

abailly committed:
Minor warnings cleanup

abailly committed:
Fix Reliability unit tests

abailly committed:
Simplify parameters passing to reliability network for parties

And do the sorting inside the function and not require a sorted vector

abailly committed:
Extract dedicated module for hydra-node's networking stack

The actual network stack used by the node is becoming more complex,
and requires pulling in dependencies which are irrelevant.

abailly committed:
Add documentation to various network modules and functions

v0d1ch committed:
Make ReliabilitySpec compile

v0d1ch committed:
Rename Msg to ReliableMsg

v0d1ch committed:
Split message resend into separate function

ghost commented:
I have added some documentation on the network components:
<img width="1440" alt="Screenshot 2023-09-20 at 08 00 23" src="https://github.com/input-output-hk/hydra/assets/79840582/0c9c5800-e883-4c62-86ac-ec5761f311a8">

abailly committed:
Ensure fixture parties are sorted as expected

We want to ensure alice, bob, carol as Party reflect the sorting
of their identifiers

v0d1ch committed:
Remove redundant imports leftover

v0d1ch committed:
Throw in case we can't find party index

pgrange committed:
Ignore malformed messages

If we receive a messagae with an unappropriate number of peers, we just
drop the message as it is quite suspicious.

pgrange committed:
Refactor: Abstract the receiption of messages from Alice

pgrange committed:
Organize tests by direction (sending vs receiving)

Can ease test exploration.

pgrange committed:
Refactor: rely on abstract aliceReceives for property

Also note that we have an issue with this property that would pass
if we drop all the messages.

pgrange committed:
FIXME note

v0d1ch committed:
More meaningfull renames

v0d1ch committed:
Introduce a log message for sent messages

- Also add a new test case

pgrange committed:
Prepare test for clearing the message queue

v0d1ch committed:
Introduce deleteSeenMessages

v0d1ch committed:
Use Map to store sent messages

v0d1ch committed:
Remove seen messages

- We record what each party has seen and remove messages seen by all
  parties.

- Use IntMap instead of Vector for storing sent messages because we need
to be able to remove old/seen indices without re-indexing.

- Use Map to keep track of the last seen message by party

- Introduce a test case to assert the withNetwork logs removal of old
  messages.

v0d1ch committed:
WIP Draft a network stress test

v0d1ch committed:
Implement a network "stress" test

We wire 2 components for alice and bob, mediated through a lossy
channel, and check bob receives all messages from alice

abailly committed:
Try fixing "stress test" by reducing list of messages size

It would be better to wait (with a timeout) to receive all broadcast messages

pgrange committed:
Waiting to see Bob trace mentioning the right amount of messages from Alice

v0d1ch committed:
All tests passing

abailly committed:
Minor cleanup of code

* Remove imports
* Remove unneeded resize
* Tabulate to show length of messages sent
* Increase ratio of dropped messages

v0d1ch committed:
Fix Reliability log test

abailly committed:
Make what the "stress test" runs clearer

abailly committed:
Rename field

abailly committed:
Updated CHANGELOG

abailly committed:
Improve Reliability logs to better legibility

abailly committed:
Remove unneeded test

This test is superseded by the stress test which covers more of it.

abailly committed:
Remove unused imports and other warnings

abailly committed:
Fix log schemas and test

pgrange committed:
Refactor: make Alice and Bob code look similar in tests

pgrange committed:
Make Bob send a few message and see what happens

v0d1ch committed:
WIP

pgrange committed:
Log when we ignore a message

pgrange committed:
This is driving me crazy:

How is it that the test fail with `[0] /= [0, 0]` but the
debug trace is `[0, 0] /= [0, 0]` ?

ch1bo commented on hydra-node/src/Hydra/Node/Network.hs:
```diff
@@ -0,0 +1,119 @@
+-- @
+--
+--          ▲                                    │
+--          │ Authenticate msg                   │ msg
```
I think you would want to refer to the actual data type here: `Authenticated msg`

ch1bo commented on hydra-node/src/Hydra/Node/Network.hs:
```diff
@@ -0,0 +1,119 @@
+-- │                                                    │
+-- └─────────▲───────────────────────────────────┼──────┘
+--           │                                   │
+--      Authenticate (ReliableMsg (Heartbeat msg))    ReliableMsg│(Heartbeat msg)
```
```suggestion
--      Authenticated (ReliableMsg (Heartbeat msg))    ReliableMsg (Heartbeat msg)
```

ch1bo commented on hydra-node/src/Hydra/Node/Network.hs:
```diff
@@ -0,0 +1,119 @@
+  -- | This node's id.
+  NodeId ->
+  -- | Produces a `NetworkComponent` that can send `msg` and consumes `Authenticated` @msg@.
+  NetworkComponent IO (Authenticated msg) msg ()
```
I think it is weird to split the callbacks of `Authenticated msg` and `Connectivity` messages via `ConnectionMessages`.

Why does this component not have a single callback handle? i.e.

```
data NetworkCallback msg m = NetworkCallback
  { onMessage :: Authenticated msg -> m ()
  , onConnectivityChange :: Connectivity -> m ()
  }
```

That would make it more clear to me in it's usage.
Maybe my observation here is that we should not be too dogmatic about making everything a `NetworkComponent`, losing `Contravariant`, but having more clarity and concreteness?

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+{-# LANGUAGE TupleSections #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- | A `Network` layer that guarantees delivery of `msg` in order even in the
```
Is "in order" delivery really important? Our protocol would be fine without it, no?

ch1bo replied:
> It's sometimes important: If you receive an AckSn before the ReqSn you will just drop it in the HeadLogic I think. 

No. The HeadLogic handles out-of-order delivery. So in a way we have more complicated than needed HeadLogic if our network layer guarantees in-order delivery? Or the network code is more complicated than needed as we actually do not need in-order transport :)

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+--
+--   * Conversely, when receiving a `Ping`, we don't update the peer's message
+--     counter but only take into account their view of /our/ counter in order
+--     to compute whether or not to resend previous messages.
```
I don't understand why the Heartbeat needs to be "going through" the reliability layer and consequently requires this special treatment? Why not have the heartbeat "below" this layer and just reliably transmit the `Data` messages?

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+import Test.QuickCheck (getPositive, listOf)
+
+data ReliableMsg msg = ReliableMsg
+  { knownMessageIds :: Vector Int
```
I'm still not sure if I understand the algorithm, but let me try to explain this field with some documentation (which I would have loved to have here) and see if it is correct:
```haskell
-- | Vector of highest known counter for each known party. Serves as announcement of
-- which messages the sender of `ReliableMsg` has seen. The individual counters have 
-- nothing to do with the `message` also included in this.
```

Did I understand it?

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+    ReliabilityFailedToFindMsg String
+  | -- | This should never happen. We should always be able to find a party
+    -- index in a list of parties.
+    ReliabilityMissingPartyIndex Party
```
If these two "should never happen". Can we avoid them?

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+-- `withHeartbeat` and `withAuthenticate` layers.
+--
+-- NOTE: better use of Vectors? We should perhaps use a `MVector` to be able to
+-- mutate in-place and not need `zipWith`
```
It appears to me that this implementation is using `Vector` already?

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+        (shouldCallback, messageAckForParty, knownAckForParty, knownAcks) <- atomically $ do
+          let messageAckForParty = acks ! partyIndex
+          knownAcks <- readTVar ackCounter
+          let knownAckForParty = knownAcks ! partyIndex
```
Can we make access into these vectors safer? i.e. any `!` in this module will raise a pure exception when out of bounds.

How are exceptions handled in the network stack? Will it tear down the hydra-node? If yes, any other party can just DoS a `hydra-node` by sending a `ReliableMsg` with empty `acks`!

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+          if messageAckForParty == knownAckForParty + 1
+            then do
+              let newAcks = constructAcks knownAcks partyIndex
+              writeTVar ackCounter newAcks
```
This is seemingly bumping the ack counter for party to be + 1. In the code here it would be much more clear if the function would actually be named like that and/or do it here? e.g. `modifyTVar ackCounter (incAckForParty partyIndex)` or so

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+    myIndex <- findPartyIndex me
+    let messageAckForUs = messageAcks ! myIndex
+    let knownAckForUs = knownAcks ! myIndex
+    when (messageAckForUs < knownAckForUs && messageAckForParty <= knownAckForParty) $ do
```
This `messageAckForParty <= knownAckForParty` decision can be taken outside of this and is even documented on line 221. I would suggest to move it there and have it read:

```
        when (messageAckForParty <= knownAckForParty) $
          resendMessages resend partyIndex sentMessages knownAcks acks
```

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+  resendMessages resend partyIndex sentMessages knownAcks messageAcks messageAckForParty knownAckForParty = do
+    myIndex <- findPartyIndex me
+    let messageAckForUs = messageAcks ! myIndex
+    let knownAckForUs = knownAcks ! myIndex
```
These vector accesses could be done right away in the beginning of `reliableCallback` and even DRY it with lines 263/264

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+              [] -> 0 -- should not happen
+              ((_, v) : _) -> v
+      sentMessages' <- readTVar sentMessages
+      if IMap.member messageReceivedByEveryone sentMessages'
```
I do not understand how the removal of messages work. The function name indicates it deletes multiple messages, but the type of `messageReceivedByEveryone` is just a single `Key` (`Int`). Maybe adding some comments would be justified here?

ch1bo commented on hydra-node/src/Hydra/Network/Reliability.hs:
```diff
@@ -0,0 +1,294 @@
+ )
+import Control.Tracer (Tracer)
+import qualified Data.IntMap as IMap
+import qualified Data.Map.Strict as Map
```
As we use a strict Map, do we want to use a strict `IntMap` as well?

ch1bo reviewed (commented):
I like that there is extensive module documentation, however I was missing comments in the actual implementation, which made reviewing harder than it could have been.

As indicated by comments in the code, the whole "layering" of this network stack is starting to leak heavily left and right. I also do think that we reached the maximum of complexity we can tackle this way and I doubt that the re-use of the `NetworkComponent` interface across these layers is helping to tackle the complexity. No need to address it in part of this PR, but I feel the urge of collapsing the whole stack into a single layer which just calls out into the individual concerns upon receiving a message / when sending a message.

On the reliability layer itself: I think the idea is fine (assuming I understood the algorithm correctly), but the implementation appears to be very fragile! In particular, access to all these vectors is often unchecked and exceptions are raised left and right if some of the assumptions are not met. As this is network code, I think we should be very defensive when dealing with received messages!

Also, I found the choice for an `IntMap` surprising. From what I can see the operations on `seenMessages` only consist of insert (on one end; on `broadcast`), sequential access (to get the missing messages which must be accessed in sequence), and delete (on the other end; when everyone has seen an index). Hence, a `Seq` might be rather what we want. Maybe not needed in this PR. What could make sense though is to encapsulate parts of the logic into smaller handles / interfaces:

```
data SentMessages m msg = SentMessages
   { insertMsg :: msg -> m ()
   , getMissing :: Int -> m [msg]
   , removeMsg :: msg -> m ()
   }
```

This should make the code a bit more readable and remove some TVar noise all over the place.

Commit:
Fixed wrong reference to type of message

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ghost commented:
@ch1bo Thanks a lot for the thorough review, as usual (and this is not flattery 😉 ). Your comments are mostly spot on, esp. w.r.t. the lack of comments documenting the actual behaviour of the code, unsafe vector accesses, and unclear naming.

On the topic of the design of the network stack and `NetworkComponent` layering, I am sympathetic to the desire to make more obvious its use, but I think the overall layering architecture should be preserved, and the leaks addressed wherever possible (it's mostly the `Reliability` layer that "leaks" a lot of needs because its logic is the most complex one, but  the other ones are pretty self contained). Layering is a pretty straightforward and common technique to separate concerns, particularly in networks (those are named "stacks" for a reason after all): Think of OSI model, TCP/IP stacks, how HTTP servers are often implemented, etc. 

While you are on vacations, I will try to propose some refactoring to the team and see how it is received and can improve the habitability of the code.

abailly committed:
Document resend algorithm

abailly committed:
Document vector of message ids

pgrange committed:
Alice and Bob where not writing fast enough

At least not fast enough compared to the time we were giving their
messages to arrive.

Sending a message once every ten seconds and expecting all the
messages to reach the peer in less than 100 seconds does not
always work.

pgrange committed:
FIX bug: do not increase local vector clock when receiving Ping

Bug was exposed by running:
```
cabal test hydra-node --test-options '-m Reliability --seed 1054015251'
```

The problem was caused by Bob increasing his local view of received
messsages from Alice from 15 to 16 when receiving a Ping from Alice
when, actually, he never received this message 16 before.

As a consequence, Alice would not resend message 16 or, when she
resends message 16, Bob would ignore it anyway as it's expecting
:x

pgrange committed:
Introduce network failure for Bob messages too

pgrange committed:
Refactor test, trying to make them more readable

ch1bo reviewed (changes requested):
I tried this today by manually running `hydra-node` from this branch in the demo and the following scenario:
* Start the demo devnet with 3 participants and open a head
* Submit a transaction via `carol` -> See `SnapshotConfirmed #1`
* Stop node of `alice`
* Submit another transaction via `carol` -> See only `TxValid`
* Start node of `alice` -> expect it to catch up and confirm the snapshot
  * Nothing happens and internally alice's node has an unrelated `AckSn` hanging in the head logic.
* Submitting yet another transaction by `carol` results in a "stuck head" situation.

According to [ADR27](https://hydra.family/head-protocol/unstable/adr/27/), this should be possible:

> Therefore, the scope of this ADR is to address only point 1. above: Ensure broadcast messages are eventually received by all peers, given the sender does not stop before.

Also, I believe we should have been explicit about this user story in the feature item #188 and we **should** have a corresponding acceptance test in scope.

pgrange committed:
Generate both Bob and Alice messages randomly

ghost commented:
@ch1bo Perhaps we want to cover that scenario, and I agree this should have been made clearer in #188 but this PR does not address it explicitly: Depending on what messages were sent by the stopping party and seen by the others, it could work or not. This PR _implicitly_ addresses point 1. which says messages can be dropped (but not for an arbitrarily long time) only if the nodes don't crash, eg. it's a pure network problem but each node keeps track of their indices.

ghost commented:
I do think we need more concrete (eg. by hand) tests, that's for sure.

ghost commented:
If we want to cover the use case you're mentioning @ch1bo I think would require persisting messages.

ch1bo commented:
> @ch1bo Perhaps we want to cover that scenario, and I agree this should have been made clearer in #188 but this PR does not address it explicitly: Depending on what messages were sent by the stopping party and seen by the others, it could work or not. This PR _implicitly_ addresses point 1. which says messages can be dropped (but not for an arbitrarily long time) only if the nodes don't crash, eg. it's a pure network problem but each node keeps track of their indices.

I have no opinion whether we do it in this PR or another time, but IMO the situation as outlined above is the core idea of #188 and what I understood we need to make possible. If that scenario in particular is not what is covered here (@pgrange thought it would be covered), then I would like to understand the user story which _is_ covered by the whole work here instead.

ghost commented:
The user story which is covered is exactly point 1. from the ADR, with the implicit provision that the behaviour in case of node "crashes" is unspecified: it could or it could not work. Perhaps there's some intermediate issue, eg. point 1.5, that covers the case where a node crashes but not while some message was in the process of being sent or in the event queue.

ghost commented:
A scenario I would expect to work is:
1. start the cluster
2. do some messages
3. cut connection between nodes
4. do more messages
5. reestablish connection
6. nodes should catch up

ghost commented:
I emulated that scenario by running the benchmarks with dropping randomly messages because actually dropping connections was more involved than what I wanted to do but I or someone else should take the time to do this properly, eg. manipulating `iptables` perhaps

ghost commented:
This is pretty explicit in the description of the PR:

> ❓ This PR only addresses part of the resilience story from https://github.com/input-output-hk/hydra/issues/188, eg. the part that deals with connection failures. Future PR should address the other issues, eg. resilience to transient node crashes

pgrange committed:
Peel one layer of useless encapsulation

pgrange committed:
WIP: FIX: do not stuck the head if two peers are lagging behind each other

If Alice is lagging behind Bob and Bob is lagging behind Alice then
nobody would resend any message to its peer.

Here we remove one condition to unlock this.

pgrange committed:
FIX: we never received messages from ourself

So we should not include ourself to the `seenMessages` map or,
otherwise, in real life, we will never garbage collect.

v0d1ch commented:
> I emulated that scenario by running the benchmarks with dropping randomly messages because actually dropping connections was more involved than what I wanted to do but I or someone else should take the time to do this properly, eg. manipulating `iptables` perhaps

We had a successful manual test with three nodes where we would kill the outgoing connectivity of one of them. Current master doesn't handle this well and is not able to catch up and produce next snapshots while this branch does exactly that.

ghost commented:
> > I emulated that scenario by running the benchmarks with dropping randomly messages because actually dropping connections was more involved than what I wanted to do but I or someone else should take the time to do this properly, eg. manipulating `iptables` perhaps
> 
> We had a successful manual test with three nodes where we would kill the outgoing connectivity of one of them. Current master doesn't handle this well and is not able to catch up and produce next snapshots while this branch does exactly that.

That's great and I think it basically means this work is done. Let's circle back tomorrow but I would like to merge this ASAP as it's been dragging on for way too long

v0d1ch commented:
> > > I emulated that scenario by running the benchmarks with dropping randomly messages because actually dropping connections was more involved than what I wanted to do but I or someone else should take the time to do this properly, eg. manipulating `iptables` perhaps
> > 
> > 
> > We had a successful manual test with three nodes where we would kill the outgoing connectivity of one of them. Current master doesn't handle this well and is not able to catch up and produce next snapshots while this branch does exactly that.
> 
> That's great and I think it basically means this work is done. Let's circle back tomorrow but I would like to merge this ASAP as it's been dragging on for way too long

I agree. Just tackled almost all @ch1bo 's concerns, there is only this one left https://github.com/input-output-hk/hydra/pull/1074#discussion_r1337823812 and perhaps to double check the documentation so let's have a merge party tomorrow.

ghost commented:
Well, the CI is still red: Some tests are failing but probably for spurious reasons, however the benchmarks are also failing which is more concerning

v0d1ch committed:
Fix logs schema for reliability messages

abailly committed:
Increase sample size for stress test

abailly committed:
Only resend messages upon receiving a Ping

This is meant to ensure we only try to resend messages whenever the peer is quiescent,
which was the original intent of using Pings in the first place in order to avoid
resending messages too often. The assumption is that disconnections and messages
drop should be few and far between in normal operations and it's therefore fine to
rely on the Ping's roundtrip time to check for peers state.

abailly committed:
[wip] Remove deleting messages

abailly committed:
Stress test peers wait for all messages to be received by both peers

Timeouts are inherently unreliable, esp. given an arbitrary and
unknown list of messages and an arbitrary ordering of actions. Tests
might fail because one of the peers stops before the other and
therefore fails to send Pings which will notify the peer it's missing
message, or fail to take into account the peer's Pings.

This commit replaces complicated timeout logic with a simple STM-based
check that _both_ peers received all the messages.

abailly committed:
Update and cleanup comments

abailly committed:
Remove map tracking peer's view

This was used for GC messages and will be rewritten later

v0d1ch committed:
WIP: Start with safe vector indexing

ffakenz committed:
Add failing test for when it cannot find the index of the own party

ffakenz committed:
Make finding our index a pure function

Also it throws an error if we do not find ourselves in the list of all parties.
This is an absurd given we included ourselves to the list before sorting.

v0d1ch committed:
Remove redundant test case

v0d1ch committed:
Remove unsafe vector access

v0d1ch committed:
Fix the logs schema

v0d1ch committed:
Remove ReliabilityException as it is not needed

v0d1ch committed:
Add timeout to the stress spec

abailly committed:
Update CHANGELOG

ghost commented:
No idea why the CI is red, going to bypass and merge it 🤷
